### ISSUE TITLE
Take mean over batch on loss twice?

### ISSUE THREAD

**Issue by elicassion** [sihyun-yu/REPA#21]
Hi @sihyun-yu 

Within the loss function we take the mean over the batch (and teacher model) once.
https://github.com/sihyun-yu/REPA/blob/b799bd9e1b5089dafff3845bae8b7dc4013e84d0/loss.py#L88

And there is another `mean()` in `train.py`
https://github.com/sihyun-yu/REPA/blob/b799bd9e1b5089dafff3845bae8b7dc4013e84d0/train.py#L306

I'm wondering whether we should only keep one of these two.

**Comment by elicassion** [comment]
> Hi, it is unintended during the refactoring. Thanks for pointing it out!

So this means the result from your paper is from the correct loss scale?

**Comment by jshang-bdai** [comment]
That makes sense! Can I simplify it to avoid for loop over the batch? Or do you have specific reasons to write in this form?

**Comment by sihyun-yu** [comment]
Sure you can! The reason why I write it in a loop form is to implement contrastive loss more easily. For cosine similarity we do not need to

**Comment by kangyiyang** [comment]
I still have this question: When the mean_flat function produces a tensor with the size [B], and considering that zs is consistently equal to 1, wouldn't it suffice to simply calculate the mean? I'm wondering why it's necessary to also divide by the batch size.

**Comment by haofuly** [comment]
> I still have this question: When the mean_flat function produces a tensor with the size [B], and considering that zs is consistently equal to 1, wouldn't it suffice to simply calculate the mean? I'm wondering why it's necessary to also divide by the batch size.

Same question. The mean_flat function produces a tensor with the size [N] rather than [N, 1]. So the proj_loss.mean() is equal to proj_loss itself.
